File: Tree/code.js
class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class Tree {
    constructor() {
        this.root = null;
        this.idx = -1;
    }

    buildTreeHelper(nodes) {
        this.idx++;
        if (nodes[this.idx] === -1) {
            return null;
        }

        const newNode = new Node(nodes[this.idx]);
        newNode.left = this.buildTreeHelper(nodes);
        newNode.right = this.buildTreeHelper(nodes);
        return newNode;
    }

    buildTree(nodes) {
        this.idx = -1; // Reset before building
        this.root = this.buildTreeHelper(nodes);
    }

    getRoot() {
        return this.root;
    }

    preorder(node) {
        if (node === null) return;
        process.stdout.write(node.data + " ");
        this.preorder(node.left);
        this.preorder(node.right);
    }

    inorder(node) {
        if (node === null) return;
        this.inorder(node.left);
        process.stdout.write(node.data + " ");
        this.inorder(node.right);
    }

    postorder(node) {
        if (node === null) return;
        this.postorder(node.left);
        this.postorder(node.right);
        process.stdout.write(node.data + " ");
    }

    levelorder(root) {
        if (root === null) return;
        let queue = [root, null];
        while (queue.length > 0) {
            let curr = queue.shift();
            if (curr === null) {
                console.log();
                if (queue.length === 0) break;
                queue.push(null);
            } else {
                process.stdout.write(curr.data + " ");
                if (curr.left) queue.push(curr.left);
                if (curr.right) queue.push(curr.right);
            }
        }
    }

    height(node) {
        if (node === null) return 0;
        return Math.max(this.height(node.left), this.height(node.right)) + 1;
    }

    countNode(node) {
        if (node === null) return 0;
        return this.countNode(node.left) + this.countNode(node.right) + 1;
    }

    sum(node) {
        if (node === null) return 0;
        return this.sum(node.left) + this.sum(node.right) + node.data;
    }
}

// -------------------- Simulating main ------------------------

const nodes = [1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1];

const tree = new Tree();
tree.buildTree(nodes);

const root = tree.getRoot();

console.log("Level Order Traversal:");
tree.levelorder(root);

console.log("Preorder:");
tree.preorder(root);
console.log("\nInorder:");
tree.inorder(root);
console.log("\nPostorder:");
tree.postorder(root);

console.log("\nSum:", tree.sum(root));
console.log("Total Nodes:", tree.countNode(root));
console.log("Height:", tree.height(root));

File: Tree/code.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
class Node
{
public:
    int data;
    Node *left;
    Node *right;
    Node(int val)
    {
        data = val;
        left = right = NULL;
    }
};

class Tree {
private:
    Node* root;
    int idx;
    Node* buildTreeHelper(vector<int>& nodes) {
        idx++;
        if(nodes[idx] == -1) return NULL;
        Node* currNode = new Node(nodes.at(idx));
        currNode->left = buildTreeHelper(nodes);
        currNode->right = buildTreeHelper(nodes);
        return currNode;
    }
public:
    Node* getRoot() {
        return root;
    }
    Tree() {
        root = NULL;
        idx = -1;
    }
    void buildTree(vector<int> nodes){
        idx = -1;
        root = buildTreeHelper(nodes);
    }
    // preorder traversal 
    void preorder(Node* node){ 
        if(root = NULL) {
            return;
        }
        cout << node->data << " ";
        preorder(node->left);
        preorder(node->right);
    }
    void inorder(Node* node){ 
        if(root = NULL) {
            return;
        }
        inorder(node->left);
        cout << node->data << " ";
        inorder(node->right);
    }
    void postorder(Node* node){ 
        if(root = NULL) {
            return;
        }
        postorder(node->left);
        postorder(node->right);
        cout << node->data << " ";
    }

    void levelOrder(Node* node) {
        if(root == NULL) {
            return;
        }
        queue<Node*> q;
        q.push(node);
        q.push(NULL);
        while(!q.empty()) {
            Node* curr = q.front();
            q.pop();
            if(curr == NULL) {
                cout << endl;
                if(q.empty()) {
                    return;
                }
                q.push(NULL);
            } else {
                cout << curr->data << " ";
                if(curr->left != NULL) {
                    q.push(curr->left);
                }
                if(curr->right != NULL) {
                    q.push(curr->right);
                }
            }
        }
    }

    int height(Node* node) {
        if(node == NULL) {
            return 0;
        }
        int leftHeight = height(node->left);
        int rightHeight = height(node->right);
        return max(leftHeight, rightHeight) + 1;
    }
    int countNode(Node* node) {
        if(node == NULL) {
            return 0;
        }
        int leftCount = countNode(node->left);
        int rightCount = countNode(node->right);
        return leftCount + rightCount + 1;
    }
    int sum(Node* node) {
        if(node == NULL) {
            return 0;
        }
        return sum(node->left) + sum(node->right) + node->data;
    }
};

int main()
{
    vector<int> nodes = {1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1};
    Tree t;
    t.buildTree(nodes);
    Node* root = t.getRoot();
    t.levelOrder(root);
    cout << t.sum(root) << endl;
    return 0;
}